torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,237 @@
+                )
+            # Iterate through all optimizer states.
+            for state_key, state_val in new_state[idx].items():
+                if state_key not in state[param_key]:
+                    raise ValueError(
+                        f"State key {state_key} not found for param {param_key}"
+                    )
+
+                src_state_val = state[param_key][state_key]
+                if isinstance(state_val, ShardedTensor):
+                    assert isinstance(src_state_val, ShardedTensor)
+                    num_shards = len(state_val.local_shards())
+                    num_new_shards = len(src_state_val.local_shards())
+                    if num_shards != num_new_shards:
+                        raise ValueError(
+                            f"Different number of shards {num_shards} vs {num_new_shards} for {param_key}/{state_key}"
+                        )
+                    for shard, src_shard in zip(
+                        state_val.local_shards(), src_state_val.local_shards()
+                    ):
+                        shard.tensor.detach().copy_(src_shard.tensor)
+                elif isinstance(state_val, torch.Tensor):
+                    assert isinstance(src_state_val, torch.Tensor)
+                    state_val.detach().copy_(src_state_val)
+                else:
+                    new_state[idx][state_key] = deepcopy(src_state_val)
+
+        # Load param_groups of state_dict
+        src_param_groups = state_dict["param_groups"]
+        new_param_groups = new_state_dict["param_groups"]
+
+        if len(new_param_groups) != len(src_param_groups):
+            raise ValueError(
+                f"Different param_groups count: {len(new_param_groups)} vs {len(src_param_groups)}"
+            )
+        src_group_map = {}
+        for group in src_param_groups:
+            param_keys = []
+            for param_key in group["params"]:
+                param_keys.append(param_key)
+            src_group_map["/".join(sorted(param_keys))] = group
+        new_group_map = {}
+        for new_group in new_param_groups:
+            param_keys = []
+            for param_key in new_group["params"]:
+                param_keys.append(self.param_keys_order[param_key])  # type: ignore[call-overload]
+            new_group_map["/".join(sorted(param_keys))] = new_group
+        for group_key, new_group in new_group_map.items():
+            if group_key not in src_group_map:
+                raise ValueError(f"Group {group_key} not found")
+            src_group = src_group_map[group_key]
+            if len(src_group) != len(new_group):
+                raise ValueError(
+                    f"Different param_group size: {len(src_group)} vs {len(new_group)}"
+                )
+            for k in src_group:
+                if k not in new_group:
+                    raise ValueError(f"Group key {k} not found for group {group_key}")
```

Comment:
Same with these other messages: It would be great if we can make explicit that the issue is the passed-in `state_dict` to `load_state_dict()`.

Reply:
sure.